protocols/mdns/src/behaviour/iface.rs

```diff
@@ -198,43 +205,40 @@ impl InterfaceState {
         params: &impl PollParameters,
     ) -> Option<(PeerId, Multiaddr, Instant)> {
         // Poll receive socket.
-        while self.recv_socket.poll_readable(cx).is_ready() {
-            match self
-                .recv_socket
-                .recv_from(&mut self.recv_buffer)
-                .now_or_never()
-            {
-                Some(Ok((len, from))) => {
-                    if let Some(packet) = MdnsPacket::new_from_bytes(&self.recv_buffer[..len], from)
-                    {
-                        self.inject_mdns_packet(packet, params);
-                    }
+        match self
+            .recv_socket
+            .poll_receive_packet(cx, &mut self.recv_buffer)
+        {
+            Poll::Ready(Some((len, from))) => {
+                if let Some(packet) = MdnsPacket::new_from_bytes(&self.recv_buffer[..len], from) {
+                    self.inject_mdns_packet(packet, params);
                 }
-                Some(Err(err)) => log::error!("Failed reading datagram: {}", err),
-                None => {}
             }
+            Poll::Ready(None) | Poll::Pending => {}
         }
+
         // Send responses.
-        while self.send_socket.poll_writable(cx).is_ready() {
-            if let Some(packet) = self.send_buffer.pop_front() {
-                match self
-                    .send_socket
-                    .send_to(&packet, SocketAddr::new(self.multicast_addr, 5353))
-                    .now_or_never()
-                {
-                    Some(Ok(_)) => log::trace!("sent packet on iface {}", self.addr),
-                    Some(Err(err)) => {
-                        log::error!("error sending packet on iface {}: {}", self.addr, err)
-                    }
-                    None => self.send_buffer.push_front(packet),
-                }
-            } else if Pin::new(&mut self.timeout).poll_next(cx).is_ready() {
-                log::trace!("sending query on iface {}", self.addr);
-                self.send_buffer.push_back(build_query());
-            } else {
-                break;
+        let mut s_buffer = VecDeque::new();
+        while let Some(packet) = self.send_buffer.pop_front() {
+            match self.send_socket.poll_send_packet(
+                cx,
+                &packet,
+                SocketAddr::new(self.multicast_addr, 5353),
+            ) {
+                Poll::Ready(_) => log::trace!("sent packet on iface {}", self.addr),
+                Poll::Pending => s_buffer.push_front(packet),
             }
         }
```

Comment:
Couldn't we already break from the loop on the first `Pending`?

Seems wasteful to empty the entire send buffer.